### PR TITLE
Fix a bug where parallely combined synchronous hooks would run twice

### PR DESCRIPTION
**Pull request by Avaq**
This fixes an issue where, if combing two parallel hooks using `ap`, where one of the hooks resolves synchronously, the continuation of the hook was being called twice.

/cc @paolobueno

### CODE REVIEW

File: index.mjs [hunk 1]
```diff
@@ -146,23 +143,34 @@ ParallelHook.prototype[$map] = function(f){
   return ParallelHook(map(f)(this.hook));
 }
 
+const crash = e => Future(() => { throw e });
+
 ParallelHook.prototype[$ap] = function(mf){
   return ParallelHook(Hook(c => {
-    let f, x, b = false;
-    const mirror = cache(never);
```
Comment (Avaq) [inline index.mjs:152]: The trick with `cache(never)` wasn't actually the issue; but I replaced this approach with a more succinct one.

File: index.mjs [hunk 2]
```diff
@@ -146,23 +143,34 @@ ParallelHook.prototype[$map] = function(f){
...
-      b = true;
-      return mirror;
-    });
-    const rx = this.hook.run(_x => {
-      x = _x;
-      if (b) return map(x => mirror.resolve(x) || x)(c(f(x)));
-      b = true;
-      return mirror;
-    });
-    return map(head)(both(rf, rx));
```
Comment (Avaq) [inline index.mjs:165]: Usage of `both`, combined with the externally-triggered resolution (previously through `mirror.resolve`, now through the capture `res` function called from `consume`), was the cause of the issue.

File: index.mjs [hunk 3]
```diff
@@ -146,23 +143,34 @@ ParallelHook.prototype[$map] = function(f){
...
+        const action = Object.create(transformation);
+        action.cancel = rx._interpret(
+          e => early(crash(e), action),
+          x => early(reject(x), action),
+          x => early(resolve(x), action),
+        );
+        return action;
+      },
+    };
+    return rf._transform(transformation);
```
Comment (Avaq) [inline index.mjs:173]: We solve the issue by using a different, custom, transformation (as opposed to `both`) which takes into account that the input Futures will resolve each other. We have to do this using Fluture's undocumented low-level API, as Fluture has not a single transformation that takes this use-case into account.

Comment (Avaq) [reply index.mjs:173]: I think that the original bug is for Fluture to solve - after all, we are able to create this issue using purely Fluture's public API. That will be a much more difficult issue to tackle, however, as it's an issue intrinsic to the design of Fluture's parallelism.

**Review by dicearr — APPROVED** [review]
Although I don't really follow what the transformation is doing.

**Pull request merged**